File: src/01-easy/unary/solution.ml
let () =
  let msg = input_line stdin in
  let bin = Buffer.create 256 in
  String.iter (fun c ->
    let n = Char.code c in
    for b = 6 downto 0 do
      Buffer.add_char bin (if n land (1 lsl b) <> 0 then '1' else '0')
    done) msg;
  let s = Buffer.contents bin in
  let len = String.length s in
  let first = ref true in
  let i = ref 0 in
  while !i < len do
    let cur = s.[!i] in
    let count = ref 0 in
    while !i < len && s.[!i] = cur do incr count; incr i done;
    if not !first then print_char ' ';
    first := false;
    Printf.printf "%s %s" (if cur = '1' then "0" else "00") (String.make !count '0')
  done;
  print_newline ()

File: src/01-easy/unary/solution.pas
program Unary;
var
  msg: string;
  bin: array[1..1400] of char;
  blen, i, j, count, n: integer;
  cur: char;
  first: boolean;
begin
  readln(msg);
  blen := 0;
  for i := 1 to Length(msg) do begin
    n := Ord(msg[i]);
    for j := 6 downto 0 do begin
      inc(blen);
      if (n shr j) and 1 = 1 then bin[blen] := '1'
      else bin[blen] := '0';
    end;
  end;
  i := 1;
  first := true;
  while i <= blen do begin
    cur := bin[i];
    count := 0;
    while (i <= blen) and (bin[i] = cur) do begin inc(count); inc(i); end;
    if not first then write(' ');
    first := false;
    if cur = '1' then write('0') else write('00');
    write(' ');
    for j := 1 to count do write('0');
  end;
  writeln;
end.
